inst/rmd/lua/number-sections.lua

```diff
@@ -0,0 +1,89 @@
+
+The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+https://github.com/atusy/lua-filters/blob/master/lua/number-sections.lua
+]]
+local attributes_free_formats = {
+  -- Improve readability of markdown formats by let them be attributes free
+  -- https://pandoc.org/MANUAL.html#general-options
+  commonmark=1,
+  commonmark_x=1,
+  gfm=1, markdown_github=1,
+  markdown=1,
+  markdown_mmd=1,
+  markdown_phpextra=1,
+  markdown_strict=1
+}
+local full_attributes = attributes_free_formats[FORMAT] == nil
+local section_number_table = {0, 0, 0, 0, 0, 0, 0, 0, 0}
+local n_section_number_table = #section_number_table
+local previous_header_level = 0
+local separator = pandoc.Space()
+if FORMAT == "docx" then -- to be consistent with Pandoc >= 2.10.1
+  separator = pandoc.Str("\t")
+end
+
+local function Meta(meta)
+  if meta.number_sections_with_attributes then
+    full_attributes = meta.number_sections_with_attributes
+  end
+end
+
+local function Header(elem)
+  -- If unnumbered
+  if (elem.classes:find("unnumbered")) then
+    if full_attributes then
+      elem.attributes["data-number"] = ""
+    end
+    return elem
+  end
+
+  -- Else
+  --- Reset and update section_number_table
+  if (elem.level < previous_header_level) then
+    for i=elem.level+1,n_section_number_table do
+      section_number_table[i] = 0
+    end
+  end
+  previous_header_level = elem.level
+  section_number_table[elem.level] = section_number_table[elem.level] + 1
+
+  --- Define section number as string
+  local section_number_string = tostring(section_number_table[elem.level])
+  if elem.level > 1 then
+    for i = elem.level-1,1,-1 do
+      section_number_string = section_number_table[i] .. "." .. section_number_string
+    end
+  end
+
+  --- Update Header element
+  table.insert(elem.content, 1, separator)
+  if full_attributes then
+    table.insert(elem.content, 1, pandoc.Span(section_number_string))
+    elem.content[1].classes = {"header-section-number"}
+    elem.attributes["data-number"] = section_number_string
```

Comment:
What is the aim of `data-number` ? 

I can't seem to make it works even with `number_sections_with_attributes: true` in my yaml. 
For which format is it interesting to have full attributes ?

Reply:
Full attributes follow the format of HTML. 

```bash
echo '# foo' | pandoc -t html --number-sections
# <h1 data-number="1" id="foo"><span class="header-section-number">1</span> foo</h1>
```

I first added this option because it would be useful when implementing `cross-ref.lua`.
However, I am not using it in my draft implemtation (<https://github.com/atusy/lua-filters/blob/master/lua/crossref.lua>).
I remove it for the sake of simplicity (at least for now).

Reply:
https://github.com/rstudio/rmarkdown/pull/1879/commits/0f2f87364aee8ed543a10ff427e63d11617f4b2e

Reply:
Just to be clear: I don't mind keeping it. It is just that when I tested it did not output what I would havec expected. 

But it is simplified for sure now ! We could add it back when required. Thanks!

Reply:
I see. Just in case if you would like to try, use

```
---
output:
  md_document:
    number_sections: true
number_sections_with_attributes: true
---

# foo
```